File: Medium/150 - EvaluateReversePolishNotation/Solution.cs
/// Question 150. Evaluate Reverse Polish Notation (https://leetcode-cn.com/problems/evaluate-reverse-polish-notation/)
/// 执行用时 : 124 ms, 在所有 C# 提交中击败了 47.85% 的用户
/// 内存消耗 : 25.4 MB, 在所有 C# 提交中击败了 25.00% 的用户

public class Solution
{
    public int EvalRPN(string[] tokens)
    {
        Stack<int> operands = new Stack<int>();

        for (int i = 0; i < tokens.Count(); i++)
        {
            string currentToken = tokens[i];
            if (currentToken.Equals("+"))
            {
                int operand1 = operands.Pop();
                int operand2 = operands.Pop();
                operands.Push(operand2 + operand1);
            }
            else if (currentToken.Equals("-"))
            {
                int operand1 = operands.Pop();
                int operand2 = operands.Pop();
                operands.Push(operand2 - operand1);
            }
            else if (currentToken.Equals("*"))
            {
                int operand1 = operands.Pop();
                int operand2 = operands.Pop();
                operands.Push(operand2 * operand1);
            }
            else if (currentToken.Equals("/"))
            {
                int operand1 = operands.Pop();
                int operand2 = operands.Pop();
                operands.Push(operand2 / operand1);
            }
            else
            {
                operands.Push(int.Parse(currentToken));
            }
        }

        return operands.Pop();
    }
}

File: Medium/150 - EvaluateReversePolishNotation/solution.ts
/// Question 150. Evaluate Reverse Polish Notation (https://leetcode-cn.com/problems/evaluate-reverse-polish-notation/)

/**
 * 执行用时：724 ms, 在所有 TypeScript 提交中击败了 5.88% 的用户
 * 内存消耗：49.4 MB, 在所有 TypeScript 提交中击败了 5.88% 的用户
 *
 * @param {string[]} tokens
 * @returns {number}
 */
function evalRPN(tokens: string[]): number {
  let numbers: number[] = [];
  const TOKENS_COUNT: number = tokens.length;

  for (let i = 0; i < TOKENS_COUNT; i++) {
    const token = tokens[i];

    if (i <= 1) {
      numbers.push(parseInt(token));
    } else {
      if (token === "+") {
        const tokenOne: number = numbers.pop();
        const tokenTwo: number = numbers.pop();
        numbers.push(tokenOne + tokenTwo);
      } else if (token === "-") {
        const tokenOne: number = numbers.pop();
        const tokenTwo: number = numbers.pop();
        numbers.push(tokenTwo - tokenOne);
      } else if (token === "*") {
        const tokenOne: number = numbers.pop();
        const tokenTwo: number = numbers.pop();
        numbers.push(tokenOne * tokenTwo);
      } else if (token === "/") {
        const tokenOne: number = numbers.pop();
        const tokenTwo: number = numbers.pop();
        const divisionResult = tokenTwo / tokenOne;
        numbers.push(
          divisionResult < 0
            ? Math.ceil(divisionResult)
            : Math.floor(divisionResult)
        );
      } else {
        numbers.push(parseInt(token));
      }
    }

    // console.log(numbers);
  }

  return numbers[0];
}
